### PR TITLE
Implement a machinery to easily access model locations on C++ and Python

### PR DESCRIPTION
**Pull request by GiulioRomualdi**
With this PR I want to propose a solution for the https://github.com/robotology/icub-models/issues/91. In 3c869f1c0d02bfa99d5a49b5c02a6e4cbb57ae25, I implemented a C++ library that will allow us to easily access the location of a model installed in by `icub-models` repo. The location of the files is automatically generated by a `cmake` machine that is in charge to set the `path` accordingly to the installation folder.

I also implemented the python bindings (with pybind11) 8d7efc78a4e9f9048546481c4a2fe2a1339d3ef6. In theory, it would be possible to use swig to support more language. However, since I've never used it I've preferred to start with something simpler.

This is an python script where I used the module. I replicated the API suggested by @traversaro in https://github.com/robotology/icub-models/issues/91#issue-912245566

```python
>>> import icub_models
>>> print(f"Models have been installed in: {icub_models.get_models_path()}")
Models have been installed in: /home/gromualdi/robot-code/robotology-superbuild/build/install/share/iCub/robots
>>> print(f"Available robots: {icub_models.get_robot_names()}")
Available robots: {'iCubErzelli02', 'iCubGazeboV3', 'iCubGenova09', 'iCubNancy01', 'iCubParis01', 'iCubLisboa01', 'iCubParis02', 'iCubGazeboV2_5_visuomanip', 'iCubGenova04_plus', 'iCubGazeboV2_5_plus', 'iCubGazeboV2_5', 'iCubGenova02', 'left_wrist_mk2', 'iCubDarmstadt01', 'iCubGenova04', 'iCubGazeboV2_6', 'iCubGenova03', 'iCubGenova01', 'iCubGazeboV2_7', 'left_hand_mk3'}
>>> for robot_name in icub_models.get_robot_names():
...     print(f"{robot_name}: {icub_models.get_model_file(robot_name)}")
... 
iCubErzelli02: /home/gromualdi/robot-code/robotology-superbuild/build/install/share/iCub/robots/iCubErzelli02/model.urdf
iCubGazeboV3: /home/gromualdi/robot-code/robotology-superbuild/build/install/share/iCub/robots/iCubGazeboV3/model.urdf
iCubGenova09: /home/gromualdi/robot-code/robotology-superbuild/build/install/share/iCub/robots/iCubGenova09/model.urdf
iCubNancy01: /home/gromualdi/robot-code/robotology-superbuild/build/install/share/iCub/robots/iCubNancy01/model.urdf
iCubParis01: /home/gromualdi/robot-code/robotology-superbuild/build/install/share/iCub/robots/iCubParis01/model.urdf
iCubLisboa01: /home/gromualdi/robot-code/robotology-superbuild/build/install/share/iCub/robots/iCubLisboa01/model.urdf
iCubParis02: /home/gromualdi/robot-code/robotology-superbuild/build/install/share/iCub/robots/iCubParis02/model.urdf
iCubGazeboV2_5_visuomanip: /home/gromualdi/robot-code/robotology-superbuild/build/install/share/iCub/robots/iCubGazeboV2_5_visuomanip/model.urdf
iCubGenova04_plus: /home/gromualdi/robot-code/robotology-superbuild/build/install/share/iCub/robots/iCubGenova04_plus/model.urdf
iCubGazeboV2_5_plus: /home/gromualdi/robot-code/robotology-superbuild/build/install/share/iCub/robots/iCubGazeboV2_5_plus/model.urdf
iCubGazeboV2_5: /home/gromualdi/robot-code/robotology-superbuild/build/install/share/iCub/robots/iCubGazeboV2_5/model.urdf
iCubGenova02: /home/gromualdi/robot-code/robotology-superbuild/build/install/share/iCub/robots/iCubGenova02/model.urdf
left_wrist_mk2: /home/gromualdi/robot-code/robotology-superbuild/build/install/share/iCub/robots/left_wrist_mk2/model.urdf
iCubDarmstadt01: /home/gromualdi/robot-code/robotology-superbuild/build/install/share/iCub/robots/iCubDarmstadt01/model.urdf
iCubGenova04: /home/gromualdi/robot-code/robotology-superbuild/build/install/share/iCub/robots/iCubGenova04/model.urdf
iCubGazeboV2_6: /home/gromualdi/robot-code/robotology-superbuild/build/install/share/iCub/robots/iCubGazeboV2_6/model.urdf
iCubGenova03: /home/gromualdi/robot-code/robotology-superbuild/build/install/share/iCub/robots/iCubGenova03/model.urdf
iCubGenova01: /home/gromualdi/robot-code/robotology-superbuild/build/install/share/iCub/robots/iCubGenova01/model.urdf
iCubGazeboV2_7: /home/gromualdi/robot-code/robotology-superbuild/build/install/share/iCub/robots/iCubGazeboV2_7/model.urdf
left_hand_mk3: /home/gromualdi/robot-code/robotology-superbuild/build/install/share/iCub/robots/left_hand_mk3/model.urdf
```

### CODE REVIEW

**Comment by GiulioRomualdi** [comment]
cc @traversaro @Nicogene

File: bindings/CMakeLists.txt
Comment (traversaro) [inline bindings/CMakeLists.txt:65]: I think that for pep-0008 the use of underscore in package names is discouraged, see https://www.python.org/dev/peps/pep-0008/#package-and-module-names, while the underscore is actuall suggested for the module name.

Comment (traversaro) [inline bindings/CMakeLists.txt:65]:
```suggestion
      file(APPEND ${CMAKE_CURRENT_BINARY_DIR}/METADATA "Name: icub-models${NEW_LINE}")
```

Comment (traversaro) [reply bindings/CMakeLists.txt:65]: Double checking with @diegoferigo .

File: bindings/CMakeLists.txt
Comment (traversaro) [inline bindings/CMakeLists.txt:70]: Note that even if we call python package name `icub-models`, the directory name here should use underscore, see https://github.com/conda-forge/opencv-feedstock/pull/300#issuecomment-1038296015 .

**Review by traversaro — APPROVED** [review]
Thanks a lot @GiulioRomualdi ! My only doubt for what regards this implementation is that by itself is not relocatable. This is not a problem as long as we just distribute binaries via apt packages (as in that case the CMAKE_INSTALL_PREFIX used to install the binaries is always the same, i.e. `/usr` or `/opt/something`) or conda (as conda handles automatically strings in the binary packages to make them relocatable).

However, if we ever want to ship this via PyPI, this may be a problem. However, I do not think this should block the PR as in any case this is a great step forward, and if necessary I think it would be trivial to replace the pybind11 Python library with a native Python library for which handling the relocatability is trivial.

File: cpp/src/iCubModels.cpp.in
Comment (traversaro) [inline cpp/src/iCubModels.cpp.in:14]: I remember that in one case conda has problem in relocating the string, I want to double check that we use the variant that does not create problems here.

Comment (traversaro) [reply cpp/src/iCubModels.cpp.in:14]: I think we should be safe here (the problematic case is the one using preprocessor macros, see https://github.com/conda-forge/libignition-gazebo-feedstock/issues/4 and https://github.com/conda/conda-build/issues/1674 . However, just to be safe I suggest a small modification:
```suggestion
    std::string retstr = "@ICUB_MODELS_MODELS_PATH@";
    return retstr;
```

**Comment by GiulioRomualdi** [comment]
Hi @traversaro, I applied the suggestions you proposed. Before merging the PR I would like to highlight:
1. I set the project version in the CMake file. The version is the same of the latest release
2. I opened the PR on devel. I don't know if it is fine or I should change the base branch to `master`

File: bindings/CMakeLists.txt
Comment (diegoferigo) [reply bindings/CMakeLists.txt]: If I interpret the PEP correctly, it states that the usage of underscores in _module_ names is discouraged. Practically, if you cannot find a single name descriptive enough, this suggestion from a very old PEP is not really followed and you can find pretty often `_`.

This being said, the module name is different from the *package* name, and this section of CMake code is about the resulting package name. In this case, it does not matter much, because underscores in the metadata are parsed as dashes by PyPI and pip. What differs is just the sdist / wheel package name. And I mean, a package with name `icub-models` containing a module stored in folder `icub_models` so that it can be imported as `import icub_models`, when packaged, produces `icub-models-....{zip|wheel}`. This bother mostly CI since you have to take care of two different names. Since packages named `icub_models` can be installed with `pip` as `pip install icub-models` regardless, I'd suggest to keep all naming consistent, using always `_`.

**Comment by traversaro** [comment]
> I opened the PR on devel. I don't know if it is fine or I should change the base branch to `master`

Devel is fine.



> I should write something in the readme

Yes, that would great otherwise people will never know of the feature! The README probably would need a revamp in general (see https://github.com/robotology/icub-models/issues/90), but I think that for this feature I think it is ok to just add two sections after https://github.com/robotology/icub-models#use-the-models-with-gazebo, something like:
* Use the models from the C++ helper library
* Use the models from the Python helper library
Where you just provide the minimal snippets of code.

File: bindings/python/CMakeLists.txt
Comment (diegoferigo) [inline bindings/python/CMakeLists.txt:23]:
```suggestion
# icub_models
```

**Review by diegoferigo — APPROVED** [review]

File: bindings/python/CMakeLists.txt
Comment (traversaro) [inline bindings/python/CMakeLists.txt:21]:
```suggestion
# Output module is:
```
(The Python module is `icub_models`, Python package is `icub-models`, according to Python guidelines)

Comment (diegoferigo) [reply bindings/python/CMakeLists.txt:21]: This is a folder containing a `__init__.py` file right? The real module is `bindings...`. Now, a package is technically any folder with a init file inside, and in terms of naming it is not the same as a pip / PyPI "package" (this is what defined in the metadata. Note that the metadata could have a name (i.e. package name `bar`) containing a python package `foo` so that you install `pip install bar` but you import with `import foo`. Not really user friendly and highly discouraged, but there are widely used packages that do not have exactly matching names). So, if I understood, calling this a package should be correct. Yeah, it's confusing :)

Comment (traversaro) [reply bindings/python/CMakeLists.txt:21]: I got it, see https://docs.python.org/3/tutorial/modules.html#packages .

File: bindings/CMakeLists.txt
Comment (xEnVrE) [inline bindings/CMakeLists.txt:49]: Similarly to https://github.com/robotology/icub-models/pull/130#discussion_r807661928, this seems to be a leftover as well.

Comment (GiulioRomualdi) [reply bindings/CMakeLists.txt:49]: You got me

**Comment by GiulioRomualdi** [comment]
@traversaro I updated the README in https://github.com/robotology/icub-models/pull/130/commits/6da40e2f84aee8c7888fa133854ff40ae0388ba1

**Comment by traversaro** [comment]
For me it is ok! Can you add a note in the CHANGELOG.md ? Thanks!

**Comment by traversaro** [comment]
Thanks!

**Pull request merged**
